flow/aggregates.py

```diff
@@ -0,0 +1,463 @@
+    :type aggregator:
+        callable
+    :param sort_by:
+        Before aggregating, sort the jobs by a given statepoint parameter.
+        The default behavior is no sorting.
+    :type sort_by:
+        str or NoneType
+    :param reverse_order:
+        States if the jobs are to be sorted in reverse order.
+        The default value is False.
+    :type reverse_order:
+        bool
+    :param select:
+        Condition for filtering individual jobs. This is passed as the
+        callable argument to `filter`.
+        The default behavior is no filtering.
+    :type select:
+        callable or NoneType
+    :param \*\*kwargs:
+        Additional information related to aggregator function.
+    """
+
+    def __init__(self, aggregator=None, sort_by=None, reverse_order=False, select=None,
+                 **kwargs):
+        if aggregator is None:
+            def aggregator(jobs):
+                return [jobs]
+
+        if not callable(aggregator):
+            raise TypeError(f"Expected callable for aggregator, got {type(aggregator)}")
+        elif sort_by is not None and not isinstance(sort_by, str):
+            raise TypeError(f"Expected string sort_by parameter, got {type(sort_by)}")
+        elif select is not None and not callable(select):
+            raise TypeError(f"Expected callable for select, got {type(select)}")
+
+        # For "non-aggregate" functions we set the Aggregate object equals to
+        # Aggregate.groupsof(1). If any other Aggregate object is associated with an
+        # operation function then mark that as a "aggregate operation" else it's a
+        # "non-aggregate" operation.
+        if sort_by is None and select is None and not reverse_order:
+            self._is_aggregate = getattr(aggregator, '_num', 0) != 1
+        else:
+            self._is_aggregate = True
+
+        self._kwargs = kwargs
+        self._aggregator = aggregator
+        self._sort_by = sort_by
+        self._reverse_order = reverse_order
+        self._select = select
+
+    @classmethod
+    def groupsof(cls, num=1, sort_by=None, reverse_order=False, select=None):
+        """signac jobs can be aggregated in groups of a number provided by an user.
+
+        By default aggregate of a single job is created.
+
+        If the number of jobs present in the project can't be divided by the number
+        provided by the user then the last aggregate will have the maximum possible
+        jobs in it.
+        For instace, if 10 jobs are present in a project and they are aggregated in groups
+        of 3 then the length of aggregates would be 3, 3, 3, 1.
+
+        The below code-block provides an example on how jobs can be aggregated in
+        groups of 2.
+
+        .. code-block:: python
+
+            @Aggregate.groupsof(num=2)
+            @FlowProject.operation
+            def foo(*jobs):
+                print(len(jobs))
+
+        :param num:
+            Maximum possible length of an aggregate.
```

Comment:
Since only the last aggregate can be a different size. I would say something like the default size of aggregates excluding the final aggregate.